boa/src/builtins/string/mod.rs

```diff
@@ -479,39 +482,90 @@ impl String {
-                        let matched = caps.get(0).expect("cannot get matched value");
-                        result = val.replace("$&", matched.as_str());
-                    }
-
-                    // Capture $1, $2, $3 etc
-                    if re.is_match(&result) {
-                        let mat_caps = re.captures(&result).unwrap();
-                        let group_str = mat_caps.get(1).unwrap().as_str();
-                        let group_int = group_str.parse::<usize>().unwrap();
-                        result = re
-                            .replace(result.as_str(), caps.get(group_int).unwrap().as_str())
-                            .to_string()
+                    let mut result = StdString::new();
+                    let mut chars = val.chars().peekable();
+
+                    let m = caps.len();
+
+                    while let Some(first) = chars.next() {
+                        if first == '$' {
+                            let second = chars.next();
+                            let second_is_digit = second.map(|ch| ch.is_digit(10)).unwrap_or(false);
+                            // we use peek so that it is still in the iterator if not used
+                            let third = if second_is_digit { chars.peek() } else { None };
+                            let third_is_digit = third.map(|ch| ch.is_digit(10)).unwrap_or(false);
+
+                            match (second, third) {
+                                (Some('$'), _) => {
+                                    // $$
+                                    result.push('$');
+                                }
+                                (Some('&'), _) => {
+                                    // $&
+                                    let matched = caps.get(0).expect("cannot get matched value");
+                                    result.push_str(matched.as_str());
+                                }
+                                (Some('`'), _) => {
+                                    // $`
+                                    let start_of_match = mat.start();
+                                    let slice = &primitive_val[..start_of_match];
+                                    result.push_str(slice);
+                                }
+                                (Some('\''), _) => {
+                                    // $'
+                                    let end_of_match = mat.end();
+                                    let slice = &primitive_val[end_of_match..];
+                                    result.push_str(slice);
+                                }
+                                (_, _) if second_is_digit && third_is_digit => {
+                                    // $nn
+                                    let tens = second.unwrap().to_digit(10).unwrap() as usize;
+                                    let units = third.unwrap().to_digit(10).unwrap() as usize;
+                                    let nn = 10 * tens + units;
```

Comment:
```suggestion
                                (Some(second), Some(third)) if second_is_digit && third_is_digit => {
                                    // $nn
                                    let tens = second.to_digit(10).unwrap() as usize;
                                    let units = third.to_digit(10).unwrap() as usize;
                                    let nn = 10 * tens + units;
```
Since we know that `second` and `third` are `char`s (not `None`) we can eliminate and `unwrap`